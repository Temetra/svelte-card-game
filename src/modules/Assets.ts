export * from "@/modules/Assets/Audio";
export * from "@/modules/Assets/CardImages";
export * from "@/modules/Assets/LoadingStatus";
export * from "@/modules/Assets/Text";