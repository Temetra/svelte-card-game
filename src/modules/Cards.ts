export * from "@/modules/Cards/Card";
export * from "@/modules/Cards/CardState";
export * from "@/modules/Cards/GetCardName";
export * from "@/modules/Cards/Rank";
export * from "@/modules/Cards/Suit";
export * from "@/modules/Cards/Scoring";
