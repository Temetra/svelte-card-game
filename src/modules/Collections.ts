interface ObjectComparer<T> {
	(first:T, second:T): number
}

export function compareArrays<T>(first: T[], second: T[], fn: ObjectComparer<T> = compareObjects) {
	return first.length == second.length 
		&& first.every((value, index) => fn(value, second[index]) == 0);
}

function compareObjects(first: any, second: any): number {
	return first === second 
		? 0 
		: first > second
			? 1
			: -1;
}