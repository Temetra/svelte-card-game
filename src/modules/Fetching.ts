export function waitFor<TParam>(ms: number, param?: TParam) : Promise<TParam> {
	return new Promise(resolve => setTimeout(() => resolve(param), ms));
}

export function checkResponse(response: Response) {
	if (!response.ok) throw { message:response.statusText, status:response.status };
	else return response;
}

export function fetchFiles<TKey extends string|number|symbol, TData>(
	files: Record<TKey, string>, 
	processResponse: {(response: Response): PromiseLike<TData>}, 
	updateProgress: {(progress: number, total: number): void} = () => {}
): Promise<Partial<Record<TKey, TData>>> {
	// Convert files to array of key/values
	let filesArr = Object.entries<string>(files);

	// Reset progress
	updateProgress(0, filesArr.length);
	let count = 0;

	// Create output dictionary
	let result: Partial<Record<TKey, TData>> = {};

	// Create array of fetch requests
	const promises: PromiseLike<any>[] = [];

	// Iterate over files
	for (let [name, source] of filesArr) {
		// Request and process data, add to result
		let request = fetch(source)
			.then(checkResponse)
			.then(processResponse)
			.then(data => result[name] = data)
			.then(() => updateProgress(++count, filesArr.length));

		// Add to array of promises
		promises.push(request);
	}

	// Wait for loading to complete then return result
	return Promise.all(promises).then(() => result);
}