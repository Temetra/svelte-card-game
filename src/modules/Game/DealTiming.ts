// Variables for controlling the timing of dealing cards
export const avgDealDelay = 400;
export const dealDelayDeviation = 25;
export const audioDelay = 0;
export const delayFrom = avgDealDelay - dealDelayDeviation - audioDelay;
export const delayTo = avgDealDelay + dealDelayDeviation - audioDelay;
export const delayFinal = avgDealDelay + dealDelayDeviation;