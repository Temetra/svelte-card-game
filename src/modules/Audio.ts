import { checkResponse } from "./Fetching";

const AudioContext = window.AudioContext || window.webkitAudioContext;
const audioCtx = new AudioContext();
const sounds: Partial<Record<string, AudioBuffer>> = {};

export function prepareAudio() {
	return Promise.all([
		loadAudio("sounds/blip.mp3"),
		loadAudio("sounds/card-hard.mp3"),
		loadAudio("sounds/card-soft.mp3")
	])
	.then(values => {
		sounds.blip = values[0];
		sounds.card = values[1];
		sounds.slide = values[2];
	});
}

function loadAudio(filename: string): Promise<AudioBuffer> {
	return fetch(filename)
		.then(checkResponse)
		.then(response => response.arrayBuffer())
		.then(arrayBuffer => audioCtx.decodeAudioData(arrayBuffer));
}

export function playSound(name: string) {
	let buffer = sounds[name];

	if (buffer) {
		const trackSource = audioCtx.createBufferSource();
		trackSource.buffer = buffer;
		trackSource.connect(audioCtx.destination);
		trackSource.start();
	}
}
