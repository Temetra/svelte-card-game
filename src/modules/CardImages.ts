import { Suit, Rank, Suits, Ranks } from "./Cards";
import { fetchFiles } from "./Fetching";
import { loading } from "@/stores/datastore";

// In-memory cache for card images
let images: Record<string, string>;

// Returns promise to preload images into bank
export function prepareCardGraphics() {
	// Create an array of images to load
	const files = {
		"CardBackground": "images/cards/CardBackground.svg",
		"Joker": "images/cards/Joker.svg"
	};
	
	// Iterate over card enums, adding a card for each combo
	for (const suit of Suits) {
		for (const rank of Ranks) {
			let name = `${Suit[suit]}${Rank[rank]}`;
			files[name] = `images/cards/${name}.svg`;
		}
	}

	// Return a promise to load images
	return fetchFiles(files, processResponse, updateProgress)
		.then(data => images = data);
}

// Get blob from response, convert to base64
async function processResponse(response: Response): Promise<string> {
	return await response.blob()
		.then(blob => new Promise<string>(resolve => {
			const reader = new FileReader();
			reader.addEventListener("error", () => resolve(null));
			reader.addEventListener("load", evt => resolve(evt.target.result as string));
			reader.readAsDataURL(blob);
		}));
}

// Update progress store
function updateProgress(progress: number, total: number) {
	loading.set({ name:"images", progress, total });
}

// Returns base64 image blob for card face background
export function getCardFace() : string {
	return images["CardBackground"];
}

// Returns base64 image blob for card face value
export function getCardValue(suit: Suit, rank: Rank) : string {
	switch (suit) {
		case Suit.Joker: return images["Joker"];
		default: return images[`${Suit[suit]}${Rank[rank]}`];
	}
}
