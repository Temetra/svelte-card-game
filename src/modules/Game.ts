import { Card, State } from "./Cards";
import { Suits, Ranks, State as CardState } from "./Cards";
import { waitFor } from "./Fetching";
import { playSound } from "./Audio";
import { controlsEnabled, deck, playerOneHand } from "@/stores/datastore";

export async function prepareGame() {
	controlsEnabled.set(true);
}

export function requestNewCards() {
	return startDealing()
		.then(dealAllCards)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(finishDealing);
}

async function startDealing() {
	// Disable button
	controlsEnabled.set(false);

	// Audio feedback
	playSound("blip");

	// Clear player hand
	playerOneHand.set([]);

	// Some sort of anim prep before dealing
	await waitFor(500);
	resetDeck();
	shuffleDeck();
}

async function dealAllCards() {
	// Array to store drawn cards
	let drawn: Card[];

	// Draw 5 cards
	deck.update(cards => {
		drawn = cards.splice(0, 5);
		return cards;
	});

	// Change state of cards
	drawn.map(card => card.state = CardState.Hidden);

	// Add card to hand
	playerOneHand.update(hand => {
		hand.push(...drawn);
		return hand;
	});
}

async function animateDealtCard() {
	const avgDelay = 400;
	const delayDeviation = 25;
	const audioDelay = 0;
	const delayFrom = avgDelay - delayDeviation - audioDelay;
	const delayTo = avgDelay + delayDeviation - audioDelay;

	// Delay before dealing
	await waitFor(randomFromRange(delayFrom, delayTo));

	// Find first hidden card in hand and change state
	playerOneHand.update(hand => {
		hand.find(x => x.state === State.Hidden).state = State.Dealing;
		return hand;
	});

	// Audio feedback
	await waitFor(audioDelay);
	playSound("card")
}

async function finishDealing() {
	// Enable button
	await waitFor(500);
	controlsEnabled.set(true);
}

function randomFromRange(min: number, max: number) : number {
	return Math.random() * (max - min) + min;
}

function resetDeck() {
	// Create new array
	const result: Card[] = [];
	const state = CardState.Default;

	// Iterate over suits and ranks, adding to array
	for (const suit of Suits) {
		for (const rank of Ranks) {
			result.push({ suit, rank, state });
		}
	}

	// Set deck store to array
	deck.set(result);
}

function shuffleDeck() {
	deck.update(array => {
		let currentIndex = array.length;
		let temporaryValue: Card;
		let randomIndex: number;
		
		while (currentIndex != 0) {
			randomIndex = Math.floor(Math.random() * currentIndex);
			currentIndex -= 1;
			temporaryValue = array[currentIndex];
			array[currentIndex] = array[randomIndex];
			array[randomIndex] = temporaryValue;
		}
		
		return array;
	});
}
