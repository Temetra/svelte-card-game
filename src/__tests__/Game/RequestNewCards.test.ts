import "../Mocks/AudioContext";
import { GameState, gameState, playerOneHand, requestNewCards } from "@/modules/Game";
import type { CardState } from "@/modules/Cards";
export{}

let states: GameState[] = [];
let unsubGameState;

let hands: CardState[][] = [];
let unsubHands;

beforeEach(() => {
	if (unsubGameState) unsubGameState();
	gameState.set(GameState.Preparing);
	states = [];
	unsubGameState = gameState.subscribe(x => states.push(x));

	if (unsubHands) unsubHands();
	playerOneHand.set([]);
	hands = [];
	unsubHands = playerOneHand.subscribe(x => hands.push(x.map(y => y.state)));
});

test("requestNewCards", async () => {
	expect(states).toStrictEqual([GameState.Preparing]);
	expect(hands).toStrictEqual([[]]);
	
	await requestNewCards();
	
	expect(states).toStrictEqual([
		GameState.Preparing,
		GameState.Dealing,
		GameState.Selecting
	]);
	
	expect(hands).toMatchSnapshot();
});
