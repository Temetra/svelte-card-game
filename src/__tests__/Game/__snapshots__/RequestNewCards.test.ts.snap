// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestNewCards 1`] = `
Array [
  Array [],
  Array [],
  Array [
    1,
    1,
    1,
    1,
    1,
  ],
  Array [
    5,
    1,
    1,
    1,
    1,
  ],
  Array [
    5,
    5,
    1,
    1,
    1,
  ],
  Array [
    5,
    5,
    5,
    1,
    1,
  ],
  Array [
    5,
    5,
    5,
    5,
    1,
  ],
  Array [
    5,
    5,
    5,
    5,
    5,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
]
`;
